File: fsharp/BWTHuffEncode.fs
namespace Benchmarks

open System
open System.Collections.Generic

[<Struct>]
type BWTResult = 
    { 
        Transformed: byte[]
        OriginalIdx: int 
    }

type HuffmanNode(frequency: int, byteVal: byte, isLeaf: bool, left: HuffmanNode, right: HuffmanNode) =
    let mutable freq = frequency
    let mutable bval = byteVal
    let mutable leaf = isLeaf
    let mutable l = left
    let mutable r = right

    new(frequency: int, byteVal: byte) = 
        HuffmanNode(frequency, byteVal, true, Unchecked.defaultof<HuffmanNode>, Unchecked.defaultof<HuffmanNode>)

    member this.Frequency = freq
    member this.ByteVal = bval
    member this.IsLeaf = leaf
    member this.Left = l
    member this.Right = r

    member this.SetLeft(node: HuffmanNode) = l <- node
    member this.SetRight(node: HuffmanNode) = r <- node
    member this.SetIsLeaf(value: bool) = leaf <- value

type HuffmanCodesStruct() =
    let codeLengths = Array.zeroCreate<int> 256
    let codes = Array.zeroCreate<int> 256

    member _.CodeLengths = codeLengths
    member _.Codes = codes

type EncodedResult = 
    { 
        Data: byte[]
        BitCount: int 
    }

type CompressedData = 
    { 
        BwtResult: BWTResult
        Frequencies: int[]
        EncodedBits: byte[]
        OriginalBitCount: int 
    }

module BWT =
    let transform (input: byte[]) : BWTResult =
        let n = input.Length
        if n = 0 then { Transformed = [||]; OriginalIdx = 0 }
        else

            let sa = Array.init n id

            let bucketLists = Array.init 256 (fun _ -> List<int>())
            for i = 0 to n - 1 do
                bucketLists.[int input.[i]].Add(i)

            let mutable pos = 0
            for bucket in bucketLists do
                for idx in bucket do
                    sa.[pos] <- idx
                    pos <- pos + 1

            if n > 1 then

                let rank = Array.zeroCreate<int> n
                let mutable currentRank = 0
                let mutable prevChar = input.[sa.[0]]

                for i = 0 to n - 1 do
                    let idx = sa.[i]
                    let currChar = input.[idx]
                    if currChar <> prevChar then
                        currentRank <- currentRank + 1
                        prevChar <- currChar
                    rank.[idx] <- currentRank

                let mutable k = 1
                while k < n do

                    let pairs = Array.zeroCreate<int * int> n
                    for i = 0 to n - 1 do
                        pairs.[i] <- (rank.[i], rank.[(i + k) % n])

                    let comparer = 
                        { new IComparer<int> with
                            member _.Compare(a, b) =
                                let (a1, a2) = pairs.[a]
                                let (b1, b2) = pairs.[b]
                                if a1 <> b1 then compare a1 b1
                                else compare a2 b2 }

                    Array.Sort(sa, comparer)

                    let newRank = Array.zeroCreate<int> n
                    newRank.[sa.[0]] <- 0

                    for i = 1 to n - 1 do
                        let prevPair = pairs.[sa.[i - 1]]
                        let currPair = pairs.[sa.[i]]
                        newRank.[sa.[i]] <- 
                            newRank.[sa.[i - 1]] + 
                            (if prevPair <> currPair then 1 else 0)

                    Array.blit newRank 0 rank 0 n
                    k <- k * 2

            let transformed = Array.zeroCreate<byte> n
            let mutable originalIdx = 0

            for i = 0 to n - 1 do
                let suffix = sa.[i]
                if suffix = 0 then
                    transformed.[i] <- input.[n - 1]
                    originalIdx <- i
                else
                    transformed.[i] <- input.[suffix - 1]

            { Transformed = transformed; OriginalIdx = originalIdx }

    let inverse (bwtResult: BWTResult) : byte[] =
        let bwt = bwtResult.Transformed
        let n = bwt.Length
        if n = 0 then [||]
        else

            let counts = Array.zeroCreate<int> 256
            for b in bwt do counts.[int b] <- counts.[int b] + 1

            let positions = Array.zeroCreate<int> 256
            let mutable total = 0
            for i = 0 to 255 do
                positions.[i] <- total
                total <- total + counts.[i]

            let next = Array.zeroCreate<int> n
            let tempCounts = Array.copy counts
            Array.fill tempCounts 0 256 0

            for i = 0 to n - 1 do
                let byteIdx = int bwt.[i]
                let pos = positions.[byteIdx] + tempCounts.[byteIdx]
                next.[pos] <- i
                tempCounts.[byteIdx] <- tempCounts.[byteIdx] + 1

            let result = Array.zeroCreate<byte> n
            let mutable idx = bwtResult.OriginalIdx

            for i = 0 to n - 1 do
                idx <- next.[idx]
                result.[i] <- bwt.[idx]

            result

module Huffman =
    let buildTree (frequencies: int[]) : HuffmanNode =
        let heap = PriorityQueue<HuffmanNode, int>()

        for i = 0 to frequencies.Length - 1 do
            if frequencies.[i] > 0 then
                heap.Enqueue(HuffmanNode(frequencies.[i], byte i), frequencies.[i])

        if heap.Count = 1 then
            let node = heap.Dequeue()
            let newNode = HuffmanNode(node.Frequency, 0uy, false, node, HuffmanNode(0, 0uy))
            newNode
        else
            while heap.Count > 1 do
                let left = heap.Dequeue()
                let right = heap.Dequeue()

                let parent = HuffmanNode(left.Frequency + right.Frequency, 0uy, false, left, right)
                heap.Enqueue(parent, parent.Frequency)

            heap.Dequeue()

    let buildCodes (node: HuffmanNode) : HuffmanCodesStruct =
        let huffmanCodes = HuffmanCodesStruct()

        let rec traverse (node: HuffmanNode) (code: int) (length: int) =
            if node.IsLeaf && node.ByteVal <> 0uy then
                let idx = int node.ByteVal
                huffmanCodes.CodeLengths.[idx] <- length
                huffmanCodes.Codes.[idx] <- code
            else
                if node.Left <> Unchecked.defaultof<HuffmanNode> then
                    traverse node.Left (code <<< 1) (length + 1)
                if node.Right <> Unchecked.defaultof<HuffmanNode> then
                    traverse node.Right ((code <<< 1) ||| 1) (length + 1)

        traverse node 0 0
        huffmanCodes

    let encode (data: byte[]) (codes: HuffmanCodesStruct) : EncodedResult =
        let result = Array.zeroCreate<byte> (data.Length * 2)
        let mutable currentByte = 0uy
        let mutable bitPos = 0
        let mutable byteIndex = 0
        let mutable totalBits = 0

        for b in data do
            let idx = int b
            let code = codes.Codes.[idx]
            let length = codes.CodeLengths.[idx]

            for i = length - 1 downto 0 do
                if (code &&& (1 <<< i)) <> 0 then
                    currentByte <- currentByte ||| (1uy <<< (7 - bitPos))

                bitPos <- bitPos + 1
                totalBits <- totalBits + 1

                if bitPos = 8 then
                    result.[byteIndex] <- currentByte
                    byteIndex <- byteIndex + 1
                    currentByte <- 0uy
                    bitPos <- 0

        if bitPos > 0 then
            result.[byteIndex] <- currentByte
            byteIndex <- byteIndex + 1

        { 
            Data = if byteIndex = result.Length then result else Array.sub result 0 byteIndex
            BitCount = totalBits 
        }

    let decode (encoded: byte[]) (root: HuffmanNode) (bitCount: int) : byte[] =
        let result = List<byte>()
        let mutable currentNode = root
        let mutable bitsProcessed = 0
        let mutable byteIndex = 0

        while bitsProcessed < bitCount && byteIndex < encoded.Length do
            let byteVal = encoded.[byteIndex]
            byteIndex <- byteIndex + 1

            for bitPos = 7 downto 0 do
                if bitsProcessed < bitCount then
                    let bit = ((byteVal >>> bitPos) &&& 1uy) = 1uy
                    bitsProcessed <- bitsProcessed + 1

                    currentNode <- if bit then currentNode.Right else currentNode.Left

                    if currentNode.IsLeaf && currentNode.ByteVal <> 0uy then
                        result.Add(currentNode.ByteVal)
                        currentNode <- root

        result.ToArray()

module Compression =
    let compress (data: byte[]) : CompressedData =
        let bwtResult = BWT.transform data

        let frequencies = Array.zeroCreate<int> 256
        for b in bwtResult.Transformed do
            frequencies.[int b] <- frequencies.[int b] + 1

        let huffmanTree = Huffman.buildTree frequencies
        let huffmanCodes = Huffman.buildCodes huffmanTree
        let encoded = Huffman.encode bwtResult.Transformed huffmanCodes

        { 
            BwtResult = bwtResult
            Frequencies = frequencies
            EncodedBits = encoded.Data
            OriginalBitCount = encoded.BitCount 
        }

    let decompress (compressed: CompressedData) : byte[] =
        let huffmanTree = Huffman.buildTree compressed.Frequencies
        let decoded = Huffman.decode compressed.EncodedBits huffmanTree compressed.OriginalBitCount
        let bwtResult = { Transformed = decoded; OriginalIdx = compressed.BwtResult.OriginalIdx }
        BWT.inverse bwtResult

type BWTHuffEncode() =
    inherit Benchmark()

    let mutable size = 0
    let mutable testData = Array.empty<byte>
    let mutable result = 0u

    let generateTestData (size: int) =
        let pattern = "ABRACADABRA"B
        Array.init size (fun i -> pattern.[i % pattern.Length])

    override this.Checksum = result

    override this.Prepare() =
        size <- int (this.ConfigVal("size"))
        testData <- generateTestData size
        result <- 0u

    override this.Run(_: int64) =
        let compressed = Compression.compress testData
        result <- result + uint32 compressed.EncodedBits.Length

File: csharp/BWTHuffEncode.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class BWTHuffEncode : Benchmark
{
    public class BWTResult
    {
        public byte[] Transformed { get; }
        public int OriginalIdx { get; }

        public BWTResult(byte[] transformed, int originalIdx)
        {
            Transformed = transformed;
            OriginalIdx = originalIdx;
        }
    }

    public BWTResult BwtTransform(byte[] input)
    {
        int n = input.Length;
        if (n == 0) return new BWTResult(Array.Empty<byte>(), 0);

        byte[] doubled = new byte[n * 2];
        Array.Copy(input, 0, doubled, 0, n);
        Array.Copy(input, 0, doubled, n, n);

        int[] sa = Enumerable.Range(0, n).ToArray();

        List<int>[] buckets = new List<int>[256];
        for (int i = 0; i < 256; i++) buckets[i] = new List<int>();

        foreach (int idx in sa) buckets[input[idx]].Add(idx);

        int pos = 0;
        for (int b = 0; b < 256; b++)
        {
            foreach (int idx in buckets[b]) sa[pos++] = idx;
        }

        if (n > 1)
        {
            int[] rank = new int[n];
            int currentRank = 0;
            byte prevChar = input[sa[0]];

            for (int i = 0; i < n; i++)
            {
                int idx = sa[i];
                byte currChar = input[idx];
                if (currChar != prevChar)
                {
                    currentRank++;
                    prevChar = currChar;
                }
                rank[idx] = currentRank;
            }

            int k = 1;
            while (k < n)
            {
                (int, int)[] pairs = new (int, int)[n];
                for (int i = 0; i < n; i++)
                {
                    pairs[i] = (rank[i], rank[(i + k) % n]);
                }

                Array.Sort(sa, (a, b) =>
                {
                    var pairA = pairs[a];
                    var pairB = pairs[b];
                    if (pairA.Item1 != pairB.Item1) return pairA.Item1.CompareTo(pairB.Item1);
                    return pairA.Item2.CompareTo(pairB.Item2);
                });

                int[] newRank = new int[n];
                newRank[sa[0]] = 0;
                for (int i = 1; i < n; i++)
                {
                    var prevPair = pairs[sa[i - 1]];
                    var currPair = pairs[sa[i]];
                    newRank[sa[i]] = newRank[sa[i - 1]] + (prevPair != currPair ? 1 : 0);
                }

                Array.Copy(newRank, rank, n);
                k *= 2;
            }
        }

        byte[] transformed = new byte[n];
        int originalIdx = 0;

        for (int i = 0; i < n; i++)
        {
            int suffix = sa[i];
            if (suffix == 0)
            {
                transformed[i] = input[n - 1];
                originalIdx = i;
            }
            else
            {
                transformed[i] = input[suffix - 1];
            }
        }

        return new BWTResult(transformed, originalIdx);
    }

    public byte[] BwtInverse(BWTResult bwtResult)
    {
        byte[] bwt = bwtResult.Transformed;
        int n = bwt.Length;
        if (n == 0) return Array.Empty<byte>();

        int[] counts = new int[256];
        foreach (byte b in bwt) counts[b]++;

        int[] positions = new int[256];
        int total = 0;
        for (int i = 0; i < 256; i++)
        {
            positions[i] = total;
            total += counts[i];
        }

        int[] next = new int[n];
        int[] tempCounts = new int[256];

        for (int i = 0; i < n; i++)
        {
            int byteIdx = bwt[i];
            int pos = positions[byteIdx] + tempCounts[byteIdx];
            next[pos] = i;
            tempCounts[byteIdx]++;
        }

        byte[] result = new byte[n];
        int idx = bwtResult.OriginalIdx;

        for (int i = 0; i < n; i++)
        {
            idx = next[idx];
            result[i] = bwt[idx];
        }

        return result;
    }

    public class HuffmanNode : IComparable<HuffmanNode>
    {
        public int Frequency { get; }
        public byte? ByteVal { get; }
        public bool IsLeaf { get; }
        public HuffmanNode? Left { get; }
        public HuffmanNode? Right { get; }

        public HuffmanNode(int frequency, byte? byteVal = null, bool isLeaf = true,
                          HuffmanNode? left = null, HuffmanNode? right = null)
        {
            Frequency = frequency;
            ByteVal = byteVal;
            IsLeaf = isLeaf;
            Left = left;
            Right = right;
        }

        public int CompareTo(HuffmanNode? other) => Frequency.CompareTo(other?.Frequency);
    }

    public HuffmanNode BuildHuffmanTree(int[] frequencies)
    {
        var heap = new PriorityQueue<HuffmanNode, int>();

        for (int i = 0; i < frequencies.Length; i++)
        {
            if (frequencies[i] > 0)
            {
                heap.Enqueue(new HuffmanNode(frequencies[i], (byte)i), frequencies[i]);
            }
        }

        if (heap.Count == 1)
        {
            var node = heap.Dequeue();
            return new HuffmanNode(node.Frequency, null, false, node, new HuffmanNode(0, 0));
        }

        while (heap.Count > 1)
        {
            var left = heap.Dequeue();
            var right = heap.Dequeue();

            var parent = new HuffmanNode(left.Frequency + right.Frequency, null, false, left, right);

            heap.Enqueue(parent, parent.Frequency);
        }

        return heap.Dequeue();
    }

    public class HuffmanCodes
    {
        public int[] CodeLengths { get; } = new int[256];
        public int[] Codes { get; } = new int[256];
    }

    public void BuildHuffmanCodes(HuffmanNode node, int code = 0, int length = 0, HuffmanCodes? huffmanCodes = null)
    {
        huffmanCodes ??= new HuffmanCodes();

        if (node.IsLeaf && node.ByteVal != 0)
        {
            int idx = node.ByteVal!.Value;
            huffmanCodes.CodeLengths[idx] = length;
            huffmanCodes.Codes[idx] = code;
        }
        else
        {
            if (node.Left != null) BuildHuffmanCodes(node.Left, code << 1, length + 1, huffmanCodes);
            if (node.Right != null) BuildHuffmanCodes(node.Right, (code << 1) | 1, length + 1, huffmanCodes);
        }
    }

    public class EncodedResult
    {
        public byte[] Data { get; }
        public int BitCount { get; }

        public EncodedResult(byte[] data, int bitCount)
        {
            Data = data;
            BitCount = bitCount;
        }
    }

    public EncodedResult HuffmanEncode(byte[] data, HuffmanCodes huffmanCodes)
    {
        byte[] result = new byte[data.Length * 2];
        byte currentByte = 0;
        int bitPos = 0;
        int byteIndex = 0;
        int totalBits = 0;

        foreach (byte b in data)
        {
            int idx = b;
            int code = huffmanCodes.Codes[idx];
            int length = huffmanCodes.CodeLengths[idx];

            for (int i = length - 1; i >= 0; i--)
            {
                if ((code & (1 << i)) != 0) currentByte |= (byte)(1 << (7 - bitPos));
                bitPos++;
                totalBits++;

                if (bitPos == 8)
                {
                    result[byteIndex++] = currentByte;
                    currentByte = 0;
                    bitPos = 0;
                }
            }
        }

        if (bitPos > 0) result[byteIndex++] = currentByte;

        return new EncodedResult(result[0..byteIndex], totalBits);
    }

    public byte[] HuffmanDecode(byte[] encoded, HuffmanNode root, int bitCount)
    {
        var result = new List<byte>(bitCount / 4 + 1);
        HuffmanNode currentNode = root;
        int bitsProcessed = 0;
        int byteIndex = 0;

        while (bitsProcessed < bitCount && byteIndex < encoded.Length)
        {
            byte byteVal = encoded[byteIndex++];

            for (int bitPos = 7; bitPos >= 0 && bitsProcessed < bitCount; bitPos--)
            {
                bool bit = ((byteVal >> bitPos) & 1) == 1;
                bitsProcessed++;

                currentNode = bit ? currentNode.Right! : currentNode.Left!;

                if (currentNode.IsLeaf && currentNode.ByteVal != 0)
                {
                    result.Add(currentNode.ByteVal!.Value);
                    currentNode = root;
                }
            }
        }

        return result.ToArray();
    }

    public class CompressedData
    {
        public BWTResult BwtResult { get; }
        public int[] Frequencies { get; }
        public byte[] EncodedBits { get; }
        public int OriginalBitCount { get; }

        public CompressedData(BWTResult bwtResult, int[] frequencies, byte[] encodedBits, int originalBitCount)
        {
            BwtResult = bwtResult;
            Frequencies = frequencies;
            EncodedBits = encodedBits;
            OriginalBitCount = originalBitCount;
        }
    }

    public CompressedData Compress(byte[] data)
    {
        BWTResult bwtResult = BwtTransform(data);

        int[] frequencies = new int[256];
        foreach (byte b in bwtResult.Transformed) frequencies[b]++;

        HuffmanNode huffmanTree = BuildHuffmanTree(frequencies);

        HuffmanCodes huffmanCodes = new HuffmanCodes();
        BuildHuffmanCodes(huffmanTree, huffmanCodes: huffmanCodes);

        EncodedResult encoded = HuffmanEncode(bwtResult.Transformed, huffmanCodes);

        return new CompressedData(bwtResult, frequencies, encoded.Data, encoded.BitCount);
    }

    public byte[] Decompress(CompressedData compressed)
    {
        HuffmanNode huffmanTree = BuildHuffmanTree(compressed.Frequencies);

        byte[] decoded = HuffmanDecode(compressed.EncodedBits, huffmanTree, compressed.OriginalBitCount);

        BWTResult bwtResult = new BWTResult(decoded, compressed.BwtResult.OriginalIdx);

        return BwtInverse(bwtResult);
    }

    public int _size;
    public byte[] _testData = Array.Empty<byte>();
    public uint _result;

    public BWTHuffEncode()
    {
        _result = 0;
        _size = (int)ConfigVal("size");
    }

    public byte[] GenerateTestData(int size)
    {
        byte[] pattern = "ABRACADABRA"u8.ToArray();
        byte[] data = new byte[size];

        for (int i = 0; i < size; i++) data[i] = pattern[i % pattern.Length];

        return data;
    }

    public override void Prepare() => _testData = GenerateTestData(_size);

    public override void Run(long IterationId)
    {
        CompressedData compressed = Compress(_testData);
        _result += (uint)compressed.EncodedBits.Length;
    }

    public override uint Checksum => _result;
}